Include instances as attributes of the relevant class

apdavison opened the pull request:
This is a proposal for how to include predefined metadata instances (e.g. controlled terms) in the Python library (cf #24). Usage example:
```python
In [1]: from openminds.latest.controlled_terms import AgeCategory

In [2]: AgeCategory.adult
Out[2]: <openminds.latest.controlled_terms.age_category.AgeCategory at 0x104eb3970>

In [3]: AgeCategory.adult.id
Out[3]: 'https://openminds.ebrains.eu/instances/ageCategory/adult'

In [4]: AgeCategory.adult.definition
Out[4]: "'Adult' categorizes the life cycle stage of an animal or human that reached sexual maturity."
```

ehennestad commented:
I don't quite understand from the jinja template whether all the instances will be embedded into the python class?

I had done something similar before in the openMINDS MATLAB package, but with suggestions from Lyuba ended up doing it differently. 

This is the example of a schema definition for AgeCategory.
https://github.com/openMetadataInitiative/openMINDS_MATLAB/blob/main/code/schemas/latest/%2Bopenminds/%2Bcontrolledterms/AgeCategory.m#L1

As you can see, it has a hidden `CONTROLLED_INSTANCES` property, and this holds the name of all known instances of this schema. In the ControlledTerm superclass there is a case in the constructor to deserialise an json ld instance if the constructor is called with an input argument which matches one of the names in the `CONTROLLED_INSTANCES` property

Example:
```
adult = openminds.controlledterms.AgeCategory('adult')

adult = 

  AgeCategory (https://openminds.ebrains.eu/controlledTerms/AgeCategory) with properties:

                     definition: "''Adult'' categorizes the life cycle stage of an animal or human that reached sexual maturity."
                    description: ""
             interlexIdentifier: "http://uri.interlex.org/base/ilx_0729043"
             knowledgeSpaceLink: ""
                           name: "adult"
    preferredOntologyIdentifier: "http://purl.obolibrary.org/obo/UBERON_0000113"
                        synonym: ["adult stage"    "post-juvenile adult"    "post-juvenile adult stage"]

  Required Properties: name

>> adult.id

ans =

    'https://openminds.ebrains.eu/instances/ageCategory/adult'
```

lzehl commented:
@ehennestad I let Andrew respond but it is not loading all instances into the class only the one of the stated type (as it is done for MATLAB as well, more or less)

lzehl commented on build.py:
```diff
@@ -7,25 +8,35 @@
 
 # Step 1 - clone central repository in main branch to get the latest sources
-clone_sources()
+###clone_sources()
```
this needs to go back no?

lzehl replied:
and extended to clone schemas as well as instances?

apdavison replied:
> this needs to go back no?

whoops! fixed

lzehl commented on pipeline/utils.py:
```diff
@@ -5,22 +5,51 @@
-
-
-class SchemaLoader(object):
+    if os.path.exists("_sources"):
```
I would rename here "_sources_schemas" for the schemas and then use "_sources_instances" for the instances. I think it should not cause too many refactoring for the schemas.

Peyman-N commented:
Python name generated for `4-(2-Hydroxyethyl)-1-piperazineEthanesulfonicAcid` is `four_(2_hydroxyethyl)_1_piperazine_ethanesulfonic_acid` which is not a valid python variable name.
`6,7-Dinitro-1,4-dihydroquinoxaline-2,3-dione` is converted to `six,7_dinitro_1,4_dihydroquinoxaline_2,3_dione`.

ehennestad commented:
Just a comment. In MATLAB I ended up doing this:


> Python name generated for `4-(2-Hydroxyethyl)-1-piperazineEthanesulfonicAcid` is `four_(2_hydroxyethyl)_1_piperazine_ethanesulfonic_acid` which is not a valid python variable name.
> `6,7-Dinitro-1,4-dihydroquinoxaline-2,3-dione` is converted to `six,7_dinitro_1,4_dihydroquinoxaline_2,3_dione`.

I also had this issue with a lot of instance names, as MATLAB is very strict about property names.

I ended up changing the constructor for controlled term classes to take the instance name as an input, for example:
`openminds.controlledterms.MolecularEntity('4-(2-Hydroxyethyl)-1-piperazineEthanesulfonicAcid')`

I think this is a better solution because you can keep using all the original names and don't have to worry about remapping them

apdavison commented:
@ehennestad and @Peyman-N I've fixed the Python name generation, but I've also added a `by_name()` class method to controlled-terms classes, similar to the solution suggested by @ehennestad. This method also uses synonyms for lookups, e.g.:

```python
In [1]: from openminds.latest.controlled_terms import MolecularEntity

In [2]: MolecularEntity.by_name("HEPES") == MolecularEntity.by_name("4-(2-hydroxyethyl)-1-piperazine ethanesulfonic acid")
Out[2]: True
```
